sources/packages/api/src/services/sin-validation/sin-validation.service.ts

```diff
@@ -1,80 +1,195 @@
+    // these fields are also the ones used later to "clone" the record if needed,
+    // as explained further along the method.
+    const existingValidation = await this.repo
+      .createQueryBuilder("sinValidation")
+      .select([
+        "sinValidation.id",
+        "sinValidation.sin",
+        "sinValidation.sinStatus",
+        "sinValidation.dateReceived",
+        "sinValidation.validSINCheck",
+        "sinValidation.dateSent",
+        "sinValidation.fileSent",
+        "sinValidation.givenNameSent",
+        "sinValidation.surnameSent",
+        "sinValidation.dobSent",
+        "sinValidation.genderSent",
+        "sinValidation.temporarySIN",
+        "sinValidation.sinExpireDate",
+        "user.id",
+      ])
+      .innerJoin("sinValidation.user", "user")
+      .where("sinValidation.id = :sinValidationId", {
+        sinValidationId: validationResponse.referenceIndex,
+      })
+      .getOne();
+
+    if (!existingValidation) {
+      throw new Error(
+        `Not able to find the SIN validation id ${validationResponse.referenceIndex} to be updated with the ESDC response.`,
+      );
+    }
+
+    const sinValidationNeverUpdated = !existingValidation.dateReceived;
+    const sinStatusChanged =
+      existingValidation.sinStatus !== validationResponse.sinCheckStatus;
+
+    // Values to be updated from the ESDC response.
+    existingValidation.dateReceived = processDate;
+    existingValidation.fileReceived = receivedFileName;
+    existingValidation.isValidSIN = isValidSIN;
+    existingValidation.sinStatus = validationResponse.sinCheckStatus;
+    existingValidation.validSINCheck = validationResponse.sinOkayFlag;
+    existingValidation.validBirthdateCheck =
+      validationResponse.birthDateOkayFlag;
+    existingValidation.validFirstNameCheck =
+      validationResponse.firstNameOkayFlag;
+    existingValidation.validLastNameCheck = validationResponse.lastNameOkayFlag;
+    existingValidation.validGenderCheck = validationResponse.genderOkayFlag;
+    existingValidation.modifier = { id: auditUserId } as User;
+
+    if (sinValidationNeverUpdated) {
+      // The SIN validation record was never updated before,
+      // update it with the information received from ESDC.
+      // This will be most common scenario.
+      operationDescription = "SIN validation record updated.";
+      const updatedRecord = await this.repo.save(existingValidation);
+      return {
+        operationDescription,
+        record: updatedRecord,
+      };
+    }
+
+    if (sinStatusChanged) {
+      // The SIN validation is already present and the received status is different from
+      // the existing one. It can happen when the status received is 2 (under review) and later
+      // a new status will be received once the review is done, what means that one request can
+      // result in more than one response. In this case, if the received record is the most updated
+      // for the student, a "cloned" record will be inserted.
```

Comment:
just wanted to know why to clone the record. why can't we update the same row status?May be we can have sync up for details

Reply:
For auditing history, when we receive a response for a record already updated, we create a new record (only if the new record is the most updated one for the student).